CONTRACTS: Fix nbsp rendering as text rather than a space

gmcew opened the pull request:
#2171 changed how contract descriptions are rendered when accessed from the terminal by removing use of `dangerouslySetInnerHTML`.

This results in the non-breaking space rendering as '`&nbsp;`' rather than '&nbsp;'. A simple solution is to revert the removal of `dangerouslySetInnerHTML` - if there's a better option I'll update this PR.

Current:
<img width="1036" height="536" alt="Image" src="https://github.com/user-attachments/assets/ac1a8d0b-a3b5-483d-94d9-e0217eda7dab" />

With use of `dangerouslySetInnerHTML`
<img width="1035" height="538" alt="Image" src="https://github.com/user-attachments/assets/e629ef50-ec60-4457-8f46-69cef41b4fbc" />

gmcew committed:
Revert use of dangerouslySetInnerHTML

Fix nbsp rendering issue

catloversg committed:
Convert string with nbsp

d0sboots commented:
Can we not just use the nbsp Unicode character directly?

gmcew committed:
Use unicode space in cct descriptions

gmcew commented:
> Can we not just use the nbsp Unicode character directly?

That looks like it works and I don't know why I didn't think of that; thank you. @catloversg thanks for sorting the workaround and your reviews.

Let me know if i missed anything else here.

EDIT: In-game view with \u0A00 for reference
<img width="1037" alt="u00a0" src="https://github.com/user-attachments/assets/92d62022-cd64-47d1-ad39-cc1c01d505fd" />
EDIT2: Triangle as run
<img width="1022" alt="runTriangle" src="https://github.com/user-attachments/assets/7cb4391f-1418-4584-9380-9354062cce90" />

EDIT2: Triangle as returned in the various forms of `getDescription`
<img width="1082" alt="getDescription" src="https://github.com/user-attachments/assets/e5564ab8-5258-4696-90b4-dd3aadde3a8f" />
<img width="1079" alt="getContractdescription" src="https://github.com/user-attachments/assets/fb88c667-b260-45ce-aade-0d165730c1e1" />

catloversg commented:
I think we should consider using a normal space character with `white-space` CSS.
```jsx
      <span key={i} style={{ whiteSpace: "pre-wrap" }}>
        {value} <br />
      </span>,
```
```js
      return [
        "Run-length encoding (RLE) is a data compression technique which encodes data as a series of runs of",
        "a repeated single character. Runs are encoded as a length, followed by the character itself. Lengths",
        "are encoded as a single ASCII digit; runs of 10 characters or more are encoded by splitting them",
        "into multiple runs.\n\n",
        "You are given the following input string:\n",
        `    ${plaintext}\n`,
        "Encode it using run-length encoding with the minimum possible output length.\n\n",
        "Examples:\n\n",
        "    aaaaabccc            ->  5a1b3c\n",
        "    aAaAaA               ->  1a1A1a1A1a1A\n",
        "    111112333            ->  511233\n",
        "    zzzzzzzzzzzzzzzzzzz  ->  9z9z1z  (or 9z8z2z, etc.)",
      ].join(" ");
```

Edit: You should also check `MinimumPathSumInATriangle` in `src\CodingContract\contracts\MinimumPathSumInATriangle.ts` and `getDescription` in `src\CodingContract\Contract.ts`. `"&nbsp;"` is used programmatically there.

gmcew committed:
Use white-space CSS and normal spaces

gmcew commented:
To clarify the edit queries first I'll put the tested screenshots in my previous comment for reference; the `MinimumPathSumTriangle` use did cause concern at the time, I did verify it still worked. The other contract of note for using `&nbsp` in more than just the description padding was `SpiralizeMatrix`; this also renders as expected, though I haven't included screenshots for that.
The `getDescription` `.replace` usage was a quick and dirty workaround about a year ago in #1525 , with hindsight it was treating the same issue that we're dealing with here - it renders fine without it, and should now be superfluous so I've removed it as part of the below change.

I can definitely see the value of not spamming unicode characters for code legibility; I've modified it as your suggestion and the results look good.
Related to your point on how the `getDescription` functions display: desciptions as returned from `getContract()` include the newline as text - but not the space - if you print the entire returned object. That happens with both the previous unicode spaces and normal space/CSS white-space approach and is a fairly niche issue.

I appreciate the help!

Triangle as run from terminal:
<img width="1025" alt="cssRun" src="https://github.com/user-attachments/assets/57620085-13fa-4fa1-8631-84c30852ef50" />

Triangle `getDescription` print:
<img width="1073" alt="cssf1" src="https://github.com/user-attachments/assets/4fa6c763-b0ba-4f8f-90f9-158105437841" />

Triangle `getContract().description` print;
<img width="1078" alt="cssf2" src="https://github.com/user-attachments/assets/712beade-5203-449d-bc56-0c6e24e50faf" />

Triangle `getContract()` print
<img width="1081" alt="cssf3" src="https://github.com/user-attachments/assets/2d183545-849e-4015-833a-495853f6825e" />
